### ISSUE TITLE
Traceback with older openssl

### ISSUE THREAD

**Issue by mimi1vx** [ronf/asyncssh#206]
With `openssl-1.1.0h`

```
>>> async def run_client():
...     async with asyncssh.connect('localhost') as con:
...         result = await conn.run('ls -al', check=True)
...         print(result.stdout, end="")
... 
>>> asyncio.get_event_loop().run_until_complete(run_client())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.7/asyncio/base_events.py", line 584, in run_until_complete
    return future.result()
  File "<stdin>", line 2, in run_client
  File "/usr/lib/python3.7/site-packages/asyncssh/misc.py", line 192, in __aenter__
    self._result = yield from self._coro
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 5724, in connect
    conn, _ = yield from create_connection(None, host, port, **kwargs)
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 5339, in create_connection
    yield from conn.wait_kex()
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 1861, in wait_kex
    yield from self._kex_waiter
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 707, in data_received
    while self._inpbuf and self._recv_handler():
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 928, in _recv_packet
    processed = handler.process_packet(pkttype, seq, packet)
  File "/usr/lib/python3.7/site-packages/asyncssh/packet.py", line 215, in process_packet
    self._packet_handlers[pkttype](self, pkttype, pktid, packet)
  File "/usr/lib/python3.7/site-packages/asyncssh/kex_ecdh.py", line 117, in _process_reply
    host_key = self._conn.validate_server_host_key(host_key_data)
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 2445, in validate_server_host_key
    self._port, key_data)
  File "/usr/lib/python3.7/site-packages/asyncssh/connection.py", line 627, in _validate_host_key
    key = decode_ssh_public_key(key_data)
  File "/usr/lib/python3.7/site-packages/asyncssh/public_key.py", line 2471, in decode_ssh_public_key
    key = handler.make_public(*key_params)
  File "/usr/lib/python3.7/site-packages/asyncssh/eddsa.py", line 70, in make_public
    public_value))
  File "/usr/lib/python3.7/site-packages/asyncssh/crypto/ed.py", line 111, in construct
    pub_key = pub_cls.from_public_bytes(pub)
  File "/usr/lib64/python3.7/site-packages/cryptography/hazmat/primitives/asymmetric/ed25519.py", line 26, in from_public_bytes
    _Reasons.UNSUPPORTED_PUBLIC_KEY_ALGORITHM
cryptography.exceptions.UnsupportedAlgorithm: ed25519 is not supported by this version of OpenSSL.
```

but `asyncssh.crypto.ed.ed25519_available` returns `False`. --> why is called python-crypthography insteded of python-libnacl

**Comment by ronf** [comment]
My apologies - I may have been too aggressive here about trying to reduce the number of external dependencies. When I saw Cryptography added support for ed25519 and ed448, I thought I could get by with only supporting that, figuring that anyone who was actually relying on ed25519 keys would likely have upgraded to a version of OpenSSL with that support in it.

It looks like I did miss checking ed25519_available in asyncssh.eddsa, though. It really should not have registered the Ed25519 public key algorithm or certificate algorithm at all on systems where ed25519_available returns false.

I can look into adding back the nacl implementation of ed25519 as a fallback when OpenSSL doesn't support it. I won't be able to offer a similar fallback for ed448, though, as nacl doesn't support that.

Longer term I'm hoping to drop calls to libnacl/libsodium entirely. Right now, the only dependency I still have on nacl is for chacha20-poly1305, and support for that is going to be available in the next release of cryptography (2.8). I already have a prototype of this working in AsyncSSH, though for now I was thinking of keeping the nacl version of chacha as a fallback.

**Comment by ronf** [comment]
Ok - libnacl support for curve25519/ed25519 is back via commit 776a87d in the "develop" branch, when it is not supported by the version of OpenSSL that cryptography is linked against. Also, this commit fixes the bug where ed25519 & ed448 were being registered even when they weren't supported on the system (leading to the traceback you saw above).

**Comment by mimi1vx** [comment]
thanks

**Comment by ronf** [comment]
This fix is now available in release 1.17.0. AsyncSSH now has support for PyCA versions of Curve25519/Ed25519, Curve448/Ed448, and Chacha20-Poly1305, if OpenSSL 1.1.1b or later is installed. However, for systems with older versions of OpenSSH, the libnacl/libsodium versions of Curve25519/Ed25519 and Chacha20-Poly1305 can be used.
